src/Pages/BasePage.php

```diff
@@ -13,47 +13,67 @@
 use HtmlFramework\Root as HtmlRoot;
 use HtmlFramework\Section as HtmlSection;
 use Utils\StringUtils;
-
+use Pages\InvalidPageException;
+
+/**
+ * This object associates `page_index.type` with:
+ *  - Some arbitrary set of display logic: doStuff()
+ *  - Some arbitrary set of display structure: printHtml()
+ *  - Some arbitrary set of data: $pageData
+ *
+ * If you ever want one Page to do something different than another or you
+ * want all the Pages to do the thing, this is the place to start.
+ */
 abstract class BasePage {
    private const TEMPLATE_PATH = 'src/templates';
+   private $ranHTMLPrint = false;
    private $pageid;
    private $pageData = [];
    private $pageIndexRows;
 
    // Name of the file we'll load in the "article" section.
    abstract protected function getPageTemplateName(): string;
-   // The "Page Slug" is what we match in the url
-   abstract protected function getPageSlug(): string;
-
    // Before we print the page we might want to do stuff.
    public function doStuff(): void {}
 
+   public function __construct(int $pageid) {
+      $this->pageid = $pageid;
+   }
+
    public function setPageData(string $index, $value): void {
+      if ($this->ranHTMLPrint) {
+         InvalidPageException::throwInvalidPageOperation("Please don't set page data after printing it. Logic will become a crazy mess!");
+      }
       $this->pageData[$index] = $value;
    }
 
-   public function matchesSlug(string $slug) {
-      return StringUtils::iMatch($this->getPageSlug(), $slug);
+   public static function matchesType(string $type) {
+      return StringUtils::iMatch(static::getPageType(), $type);
    }
 
    public function printHtml(): void {
+      $this->ranHTMLPrint = true;
       $htmlHead = HtmlHead::fromValues($this->getPageTitle());
       $htmlHeader = HtmlHeader::fromValues($this->getPageHeader());
       $htmlNav = HtmlNav::fromValues();
-      $htmlArticle = HtmlArticle::fromValues($this->getPageid(), $this->getPageTemplatePath(), $this->pageData, $this->getMainArticle());
+      $htmlArticle = HtmlArticle::fromValues(static::getPageType(), $this->getPageid(), $this->getPageTemplatePath(), $this->pageData, $this->getMainArticle());
```

Comment:
This is starting to get ugly but I think it's the last new parameter we'll need.